firedrake/preconditioners/fdm.py

```diff
@@ -0,0 +1,1134 @@
+        A, P = pc.getOperators()
+
+        # Read options
+        prefix = pc.getOptionsPrefix()
+        options_prefix = prefix + self._prefix
+        # opts = PETSc.Options(options_prefix)
+
+        dm = pc.getDM()
+        V = get_function_space(dm)
+        use_tensorproduct, N, ndim, family, _ = tensor_product_space_query(V)
+
+        if not use_tensorproduct or (family < {"RTCE", "NCE"}):
+            raise ValueError("FDMPC does not support the element %s" % V.ufl_element())
+
+        needs_interior_facet = not (family <= {"Q", "Lagrange"})
+        needs_hdiv = family < {"RTCF", "NCF"}
+        Nq = 2*N+1  # quadrature degree, gives exact interval stiffness matrices for constant coefficients
+
+        self.mesh = V.ufl_domain()
+        self.uf = firedrake.Function(V)
+        self.uc = firedrake.Function(V)
+        self.cell_node_map = get_permuted_map(V)
+
+        # Get Jacobian form and bcs
+        solverctx = get_appctx(dm)
+        self.J = solverctx.J
+        self.bcs = solverctx.bcs_J
+
+        if len(self.bcs) > 0:
+            self.bc_nodes = numpy.unique(numpy.concatenate([bcdofs(bc, ghost=False) for bc in self.bcs]))
+        else:
+            self.bc_nodes = numpy.empty(0, dtype=PETSc.IntType)
+
+        bcflags = self.get_bc_flags(self.bcs, self.J)
+
+        self.weight = self.multiplicity(V)
+        with self.weight.dat.vec as w:
+            w.reciprocal()
+
+        # Get the interior penalty parameter from the appctx
+        appctx = self.get_appctx(pc)
+        eta = float(appctx.get("eta", (N+1)*(N+ndim)))
+
+        # Get the FDM transfer kernels (restriction and prolongation)
+        # Afdm = sparse interval mass and stiffness matrices for each direction
+        # Dfdm = tabulation of normal derivative of the FDM basis at the boundary for each direction
+        Afdm, Dfdm, self.restrict_kernel, self.prolong_kernel = self.assemble_matfree(V, N, Nq, eta, needs_interior_facet, needs_hdiv)
+
+        # Get coefficients w.r.t. the reference coordinates
+        # we may use a lower quadrature degree, but using Nq is not so expensive
+        coefficients, self.assembly_callables = self.assemble_coef(self.J, Nq, discard_mixed=True, cell_average=True, needs_hdiv=needs_hdiv)
+
+        # Set arbitrary non-zero coefficients for preallocation
+        for coef in coefficients.values():
+            with coef.dat.vec as cvec:
+                cvec.set(1.0E0)
+
+        # Preallocate by calling the assembly routine on a PREALLOCATOR Mat
+        prealloc = PETSc.Mat().create(comm=A.comm)
+        prealloc.setType(PETSc.Mat.Type.PREALLOCATOR)
+        prealloc.setSizes(A.getSizes())
+        prealloc.setUp()
+        self.assemble_kron(prealloc, V, coefficients, Afdm, Dfdm, eta, bcflags, needs_hdiv)
+        nnz = get_preallocation(prealloc, V.value_size * V.dof_dset.set.size)
+
+        self.Pmat = PETSc.Mat().createAIJ(A.getSizes(), nnz=nnz, comm=A.comm)
+        self.Pmat.setBlockSize(V.value_size)
```

Comment:
Does firedrake use `BAIJ`? I thought matrices for vector fields were always `AIJ`.

Reply:
It can do, but it appears not to be the default.
```
In [12]: V = VectorFunctionSpace(mesh, "P", 1)

In [13]: u = TrialFunction(V)

In [14]: v = TestFunction(V)

In [15]: a = inner(u, v)*dx

In [16]: A = assemble(a)
A.petscmat.getType()

In [17]: A.petscmat.getType()
Out[17]: 'seqaij'

In [18]: A = assemble(a, mat_type="baij")

In [19]: A.petscmat.getType()
Out[19]: 'seqbaij'
```

I'm happy to leave this is as is.

Reply:
BAIJ does not seem to be very compatible with CHOLMOD
```
[0]PETSC ERROR: MatSolverType cholmod does not support matrix type seqbaij
```
If I use PETSc instead, then nested dissection is not available:
```
[0]PETSC ERROR: Matrix reordering is not supported
```
So I'm also happy to leave it as it is.